extract_missing_links.py

```diff
@@ -0,0 +1,386 @@
+#!/usr/bin/env python3
+
+import difflib
+from io import BytesIO
+import itertools
+import json
+import re
+import string
+import subprocess
+import tempfile
+from urllib.parse import urljoin, urlparse
+from urllib.request import urlopen
+from zipfile import ZipFile
+
+from bs4 import BeautifulSoup
+from pathlib import Path
+import sys
+
+NEW_VERSION = "3.8"
+OLD_VERSION = "2.7"
+
+PYTHON2_VERSIONS = ["2.0", "2.1", "2.2", "2.3", "2.4", "2.5", "2.6", "2.7"]
+
+JS_FILE_WITH_SPECIAL_CASES = "./special-cases.js"  # has to have leading "./"
+if not Path(JS_FILE_WITH_SPECIAL_CASES).is_file():
+    raise SystemExit(f"Couldn't find file {JS_FILE_WITH_SPECIAL_CASES}")
+DOWNLOAD_DOCS_TO = Path("docs.python.org")
+OUTPUT_DIRECTORY = Path("links")
+
+ARCHIVE_URLS = {
+    "2.0": "https://www.python.org/ftp/python/doc/2.0.1/html-2.0.1.zip",
+    "2.1": "https://www.python.org/ftp/python/doc/2.1.3/html-2.1.3.zip",
+    "2.2": "https://www.python.org/ftp/python/doc/2.2.3/html-2.2.3.zip",
+    "2.3": "https://www.python.org/ftp/python/doc/2.3.5/html-2.3.5.zip",
+    "2.4": "https://www.python.org/ftp/python/doc/2.4.4/html-2.4.4.zip",
+    "2.5": "https://www.python.org/ftp/python/doc/2.5.4/html-2.5.4.zip",
+    "2.6": "http://docs.python.org/ftp/python/doc/2.6.9/python-2.6.9-docs-html.zip",
+    "2.7": "https://docs.python.org/2/archives/python-2.7.18-docs-html.zip",
+    "3.8": "https://docs.python.org/3/archives/python-3.8.3-docs-html.zip",
+}
+# The encoding was changed from Windows-1252 in Python 2.5
+ENCODING_CHANGED_VERSION = "2.5"
+
+# HTML ids of non-documentation elements
+USELESS_IDS = {
+    "documentation_options",
+    "outdated-warning",
+    "searchlabel",
+    "searchbox",
+    # Python 2.5
+    "top-navigation-panel",
+}
+
+
+def version_to_dir(version):
+    return DOWNLOAD_DOCS_TO / ("3" if version == NEW_VERSION else version)
+
+
+# Download a local copy of the documentation to ./docs.python.org/{version}/
+for version, url in ARCHIVE_URLS.items():
+    output_dir = version_to_dir(version)
+    if output_dir.is_dir():
+        continue
+
+    DOWNLOAD_DOCS_TO.mkdir(parents=True, exist_ok=True)
+    zip_file = output_dir.parent / (output_dir.name + ".zip")
+    with tempfile.TemporaryDirectory() as temp_dir:
+        temp_path = Path(temp_dir) / "python_docs"
+        # print(f"Downloading {url} to {temp_path}", file=sys.stderr)
+        with urlopen(url) as f:
+            with ZipFile(BytesIO(f.read())) as docs_zip:
+                docs_zip.extractall(path=temp_path)
+
+        # Some archives extract to a single directory, the older ones don't
+        unzipped_contents = list(temp_path.iterdir())
+        if len(unzipped_contents) == 1:
+            temp_path = unzipped_contents[0]
+        temp_path.rename(output_dir)
```

Comment:
When not running the script inside the same filesystem as temp location, this line produces the following exception:
```
OSError: [Errno 18] Invalid cross-device link: '/tmp/tmpcnr8gfgx/python_docs' -> 'docs.python.org/2.0'
```
```suggestion
        shutil.move(temp_path, output_dir)
```